cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java

```diff
@@ -89,26 +98,35 @@
 import java.util.NoSuchElementException;
 import java.util.Set;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
 import javax.annotation.Nullable;
 
 /**
  * Implementation of the Store that ultimately places data into MetaDataTable.
  */
 public class DefaultStore implements Store {
-
   private static final Logger LOG = LoggerFactory.getLogger(DefaultStore.class);
   private static final DatasetId APP_META_INSTANCE_ID = NamespaceId.SYSTEM.dataset(Constants.AppMetaStore.TABLE);
+  private static final byte[] APP_VERSION_UPGRADE_KEY = Bytes.toBytes("version.default.store");
+  private static final String NAME = DefaultStore.class.getSimpleName();
+
   // mds is specific for metadata, we do not want to add workflow stats related information to the mds,
   // as it is not specifically metadata
   private static final DatasetId WORKFLOW_STATS_INSTANCE_ID = NamespaceId.SYSTEM.dataset("workflow.stats");
   private static final Gson GSON = new Gson();
   private static final Map<String, String> EMPTY_STRING_MAP = ImmutableMap.of();
   private static final Type STRING_MAP_TYPE = new TypeToken<Map<String, String>>() { }.getType();
+  private static final AtomicBoolean appMetaStoreInitialized = new AtomicBoolean(false);
+  private static final AtomicBoolean upgradeComplete = new AtomicBoolean(false);
+  private static final AtomicBoolean cacheLoaderInitialized = new AtomicBoolean(false);
 
   private final CConfiguration configuration;
   private final DatasetFramework dsFramework;
   private final Transactional transactional;
 
+  private static LoadingCache<byte[], Boolean> upgradeCacheLoader;
```

Comment:
can directly initialize it here instead of in the constructor, and no need of `cacheLoaderInitialized`